lute/io/src/io.cpp

```diff
@@ -0,0 +1,141 @@
+#include "lute/io.h"
+#include "Luau/Variant.h"
+#include "lute/runtime.h"
+
+#include <uv.h>
+#include <memory>
+
+#include "lua.h"
+#include "lualib.h"
+
+
+namespace io
+{
+
+struct IOHandle
+{
+    Luau::Variant<uv_pipe_t, uv_tty_t> streamVariant;
+    uv_loop_t* loop = nullptr;
+    ResumeToken resumeToken;
+    std::shared_ptr<IOHandle> self;
+    uv_buf_t *buf;
+
+    ~IOHandle() {
+        if (buf) {
+            free(buf->base);
+            buf = nullptr;
+        }
+    }
+
+    void closeHandles()
+    {
+        auto closeCb = [](uv_handle_t* handle)
+        {
+            IOHandle* ioh = static_cast<IOHandle*>(handle->data);
+            ioh->self.reset();
+        };
+
+        uv_stream_t *stream = getStream();
+        uv_read_stop(stream);
+        uv_close((uv_handle_t*)stream, closeCb);
+    }
+
+    uv_stream_t* getStream() {
+        if (auto* tty = streamVariant.get_if<uv_tty_t>())
+            return (uv_stream_t*)tty;
+        if (auto* pipe = streamVariant.get_if<uv_pipe_t>())
+            return (uv_stream_t*)pipe;
+        return nullptr;
+    }
+};
+
+static void allocBuffer(uv_handle_t* handle, size_t suggestedSize, uv_buf_t* buf)
+{
+    buf->base = (char*)malloc(suggestedSize);
+    buf->len = buf->base ? suggestedSize : 0;
```

Comment:
We shouldn't do allocations this way. We're doing C++, even if the libuv APIs are C. If the IOHandle owns the buffer, then we should have the backing allocation owned by IOHandle, e.g. it should have a `std::vector<char>` or a `std::unique_ptr<std::vector<char>>` and calls consuming the memory as a `uv_buf_t` should temporarily construct the `uv_buf_t` in reference to that backing allocation. There should not be `malloc` and `free` in Lute's code. If it's anywhere else, that is a bug to be fixed as well.